core/src/main/java/org/elasticsearch/tasks/TaskManager.java

```diff
@@ -39,6 +41,8 @@
 
     private final AtomicLong taskIdGenerator = new AtomicLong();
 
+    private final Collection<TaskManagerListener> listeners = new CopyOnWriteArrayList<>();
```

Comment:
Would it be possible to make this a unmodifiable map and pass in all the listeners as ctor args? I really think this would simplify code dramatically also future wise? Then we can also drop the remove method?